gobblin-data-management/src/main/java/gobblin/data/management/retention/action/RetentionAction.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.data.management.retention.action;
+
+import java.io.IOException;
+import java.util.List;
+
+import org.apache.hadoop.fs.FileSystem;
+
+import com.typesafe.config.Config;
+
+import gobblin.data.management.dataset.Dataset;
+import gobblin.data.management.policy.VersionSelectionPolicy;
+import gobblin.data.management.version.DatasetVersion;
+import gobblin.data.management.version.finder.VersionFinder;
+import gobblin.dataset.DatasetsFinder;
+
+
+/**
+ * An abstraction to perform a retention action for a subset of {@link DatasetVersion}s.
+ * A few kinds of actions are deletion, access control, encryption, archival etc.
+ */
+public abstract class RetentionAction {
+
+  protected final FileSystem fs;
+
+  public RetentionAction(Config actionConfig, FileSystem fs) {
+    this.fs = fs;
+  }
+
+  /**
+   * Execute the action on all {@link DatasetVersion}s or a subset of {@link DatasetVersion}s. Each {@link Dataset}
+   * uses the {@link VersionFinder} to find all the {@link DatasetVersion}s and calls this method to perform the necessary
+   * action on those {@link DatasetVersion}s
+   * <p>
+   * <b>Note</b> Any kind of {@link VersionSelectionPolicy} has <b>NOT</b> been applied to the list of {@link DatasetVersion}s
```

Comment:
Each action can have its own `SelectionPolicy` of several `SelectionPolicies`. This give more flexibility for the the action itself to decide what to do with all the versions.

Another approach would have been that the dataset calls `RetentionAction.getSelectionPolicy`, filters versions using this policy and then calls `RetentionAction.execute` on them. This does not work well with `MultiAccessControlAction` as there it has multiple `SelectionPolicy` and permission combination.